fish-rust/src/builtins/abbr.rs

```diff
@@ -0,0 +1,595 @@
+pub use widestring::Utf32String as WString;
+
+const CMD: &wstr = L!("abbr");
+
+#[derive(Default, Debug)]
+struct Options {
+    add: bool,
+    rename: bool,
+    show: bool,
+    list: bool,
+    erase: bool,
+    query: bool,
+    function: Option<WString>,
+    regex_pattern: Option<WString>,
+    position: Option<Position>,
+    set_cursor_marker: Option<WString>,
+    args: Vec<WString>,
+}
+
+impl Options {
+    fn validate(&mut self, streams: &mut io_streams_t) -> bool {
+        // Duplicate options?
+        let mut cmds = vec![];
+        if self.add {
+            cmds.push(L!("add"))
+        };
+        if self.rename {
+            cmds.push(L!("rename"))
+        };
+        if self.show {
+            cmds.push(L!("show"))
+        };
+        if self.list {
+            cmds.push(L!("list"))
+        };
+        if self.erase {
+            cmds.push(L!("erase"))
+        };
+        if self.query {
+            cmds.push(L!("query"))
+        };
+
+        if cmds.len() > 1 {
+            streams.err.append(wgettext_fmt!(
+                "%ls: Cannot combine options %ls\n",
+                CMD,
+                join(&cmds, L!(", "))
+            ));
+            return false;
+        }
+
+        // If run with no options, treat it like --add if we have arguments,
+        // or --show if we do not have any arguments.
+        if cmds.is_empty() {
+            self.show = self.args.is_empty();
+            self.add = !self.args.is_empty();
+        }
+
+        if !self.add && self.position.is_some() {
+            streams.err.append(wgettext_fmt!(
+                "%ls: --position option requires --add\n",
+                CMD
+            ));
+            return false;
+        }
+        if !self.add && self.regex_pattern.is_some() {
+            streams
+                .err
+                .append(wgettext_fmt!("%ls: --regex option requires --add\n", CMD));
+            return false;
+        }
+        if !self.add && self.function.is_some() {
+            streams.err.append(wgettext_fmt!(
+                "%ls: --function option requires --add\n",
+                CMD
+            ));
+            return false;
+        }
+        if !self.add && self.set_cursor_marker.is_some() {
+            streams.err.append(wgettext_fmt!(
+                "%ls: --set-cursor option requires --add\n",
+                CMD
+            ));
+            return false;
+        }
+        if self
+            .set_cursor_marker
+            .as_ref()
+            .map(|m| m.is_empty())
+            .unwrap_or(false)
+        {
+            streams.err.append(wgettext_fmt!(
+                "%ls: --set-cursor argument cannot be empty\n",
+                CMD
+            ));
+            return false;
+        }
+
+        return true;
+    }
+}
+
+fn join(list: &[&wstr], sep: &wstr) -> WString {
+    let mut result = WString::new();
+    for s in list {
+        result.push_utfstr(s);
+        result.push_utfstr(sep);
```

Comment:
This looks wrong, we don't want to add the separator after the last element.
Can we not use [slice::join](https://doc.rust-lang.org/std/primitive.slice.html#method.join)?

Reply:
`[&wstr]` Does not implement `Join` trait